From c335be4beec39cc5617ee7d50c8e7ca5ededac59 Mon Sep 17 00:00:00 2001
From: "tristan.ramseyer" <tristan.ramseyer@stromerbike.com>
Date: Tue, 3 Sep 2024 15:10:31 +0200
Subject: [PATCH] clk fix

Patch by Christian Duenki <christian.duenki@escatec.com>
Upstream-Status: Inappropriate [stromer-specific]

Signed-off-by: tristan.ramseyer <tristan.ramseyer@stromerbike.com>
---
diff --git a/arch/arm/mach-imx/mach-imx6ul.c b/arch/arm/mach-imx/mach-imx6ul.c
index cb6d29c2bb5f..880e40b320c3 100644
--- a/arch/arm/mach-imx/mach-imx6ul.c
+++ b/arch/arm/mach-imx/mach-imx6ul.c
@@ -3,6 +3,7 @@
  * Copyright (C) 2015 Freescale Semiconductor, Inc.
  */
 #include <linux/irqchip.h>
+#include <linux/of_address.h>
 #include <linux/of_platform.h>
 #include <asm/mach/arch.h>
 
@@ -12,12 +13,45 @@
 
 static void __init imx6ul_init_machine(void)
 {
+	void __iomem *iomux;
+	struct device_node *np;
+
 	imx_print_silicon_rev(cpu_is_imx6ull() ? "i.MX6ULL" : "i.MX6UL",
 		imx_get_soc_revision());
 
 	of_platform_default_populate(NULL, NULL, NULL);
 	imx_anatop_init();
 	imx6ul_pm_init();
+
+	np = of_find_compatible_node(NULL,NULL,"fsl,imx6ul-iomuxc");
+	iomux = of_iomap(np, 0);
+
+	/* Set daisy-chain register to avoid uart7 of getting interrupts from rts */
+	writel_relaxed(0x2,iomux+0x650);
+
+	/* Overwriting mux selection for GPIO1_IO00 (Cadence) */
+	writel_relaxed(0x5,iomux+0x05C);
+
+	/* Overwriting mux selection for GPIO2_IO16 (MPIO_1) */
+	writel_relaxed(0x5,iomux+0x1bc);
+
+	/* Overwriting mux selection for GPIO4_IO16 (MPIO_2) */
+	writel_relaxed(0x5,iomux+0x01b8);
+
+	/* Overwriting mux selection for GPIO1_IO18 (MPIO1_DIR) */
+	writel_relaxed(0x5,iomux+0x008c);
+
+	/* Overwriting mux selection for GPIO1_IO19 (MPIO2_DIR) */
+	writel_relaxed(0x5,iomux+0x0090);
+
+	/* Overwriting mux selection for GPIO4_IO18 (I2C4_ON) */
+	writel_relaxed(0x5,iomux+0x01D8);
+
+	/* Overwriting mux selection for GPIO2_IO19 (I2C4_PU) */
+	writel_relaxed(0x5,iomux+0x01C8);
+
+	/* Overwriting mux selection for GPIO2_IO17 (BRAKE_PUPD) */
+	writel_relaxed(0x5,iomux+0x1C0);
 }
 
 static void __init imx6ul_init_irq(void)
-- 
2.34.1

