diff --git a/common/state/backend_bucket_direct.c b/common/state/backend_bucket_direct.c
index 117cdfb46c..77a5ce4781 100644
--- a/common/state/backend_bucket_direct.c
+++ b/common/state/backend_bucket_direct.c
@@ -45,6 +45,17 @@ static inline struct state_backend_storage_bucket_direct
 			    bucket);
 }
 
+static void state_backend_bucket_direct_adjust(struct state_backend_storage_bucket
+					    *bucket, off_t offset,
+					    ssize_t max_size)
+{
+	struct state_backend_storage_bucket_direct *direct =
+	    get_bucket_direct(bucket);
+
+	direct->offset = offset;
+	direct->max_size = max_size;
+}
+
 static int state_backend_bucket_direct_read(struct state_backend_storage_bucket
 					    *bucket, void ** buf_out,
 					    ssize_t * len_out)
@@ -183,6 +194,7 @@ int state_backend_bucket_direct_create(struct device_d *dev, const char *path,
 
 	direct->bucket.read = state_backend_bucket_direct_read;
 	direct->bucket.write = state_backend_bucket_direct_write;
+	direct->bucket.adjust = state_backend_bucket_direct_adjust;
 	direct->bucket.free = state_backend_bucket_direct_free;
 	*bucket = &direct->bucket;
 
diff --git a/common/state/backend_format_raw.c b/common/state/backend_format_raw.c
index 1fecdeb9cf..914fb1021d 100644
--- a/common/state/backend_format_raw.c
+++ b/common/state/backend_format_raw.c
@@ -115,10 +115,11 @@ static int backend_format_raw_verify(struct state_backend_format *format,
 
 	header = (struct backend_raw_header *)buf;
 	crc = crc32(0, header, sizeof(*header) - sizeof(uint32_t));
-	if (crc != header->header_crc)
-		return dev_err_state_init(backend_raw->dev, header->header_crc ? -EINVAL : -ENOMEDIUM,
-			"header crc in raw format, calculated 0x%08x, found 0x%08x\n",
+	if (crc != header->header_crc) {
+		dev_notice(backend_raw->dev, "Error, invalid header crc in raw format, calculated 0x%08x, found 0x%08x\n",
 			crc, header->header_crc);
+		return -EINVAL;
+	}
 
 	if (magic && magic != header->magic) {
 		dev_err(backend_raw->dev, "Error, invalid magic in raw format 0x%08x, should be 0x%08x\n",
@@ -179,6 +180,7 @@ static int backend_format_raw_unpack(struct state_backend_format *format,
 				     ssize_t len)
 {
 	struct state_variable *sv;
+	struct state_variable *sv_tmp;
 	const struct backend_raw_header *header;
 	const void *data;
 	struct state_backend_format_raw *backend_raw = get_format_raw(format);
@@ -187,11 +189,11 @@ static int backend_format_raw_unpack(struct state_backend_format *format,
 	header = (const struct backend_raw_header *)buf;
 	data = buf + sizeof(*header);
 
-	list_for_each_entry(sv, &state->variables, list) {
+	list_for_each_entry_safe(sv, sv_tmp, &state->variables, list) {
 		if (sv->start + sv->size > header->data_len) {
-			dev_err(backend_raw->dev, "State variable ends behind valid data, %s\n",
+			dev_notice(backend_raw->dev, "State variable ends behind valid data, %s\n",
 				sv->name);
-			ret = -ENOSPC;
+			list_del(&sv->list);
 			continue;
 		}
 		memcpy(sv->raw, data + sv->start, sv->size);
diff --git a/common/state/backend_storage.c b/common/state/backend_storage.c
index c55d22e37f..b41b05a37d 100644
--- a/common/state/backend_storage.c
+++ b/common/state/backend_storage.c
@@ -49,6 +49,40 @@
 
 static const unsigned int min_buckets_written = 1;
 
+/*
+ * The early barebox versions for the imx6ul-medusa did have a smaller
+ * backend-stridesize of 0x53 as compared to the newer ones with 0x70 because
+ * the following variables did not exist:
+ * - update
+ * - barebox
+ * - uart8
+ * - label
+ * 
+ * To avoid maintaining a separate barebox variant for imx6ul-medusa machines
+ * which where provisioned with those early barebox versions:
+ * - Upon detection of broken buckets after read & verify, a retrial is
+ *   attempted with the legacy value of the backend-stridesize and on success
+ *   the storage backend-stridesize adjusted.
+ * - In case the storage backend-stridesize value equals the legacy value,
+ *   the newer variables are not added during the fixup so that barebox-state
+ *   in userpace does not run into problems.
+ * - State variables which end behind the valid data are removed from the
+ *   internal list with variables to avoid being written and currupt data.
+ *   Allocated memory is not freed up to be able to use the state variables
+ *   in command line without causing a crash.
+ *
+ * Remarks: For simplicity, the barebox internal devicetree and state variables
+ *          (for environments and the command line) are not adjusted after
+ *          adjusting the storage backend-stridesize. Also, the following
+ *          will be printed during barebox startup for imx6ul-medusa machines
+ *          which where provisioned with those early barebox versions:
+ *          - state: State variable ends behind valid data, update
+ *          - state: State variable ends behind valid data, barebox
+ *          - state: State variable ends behind valid data, uart8
+ *          - state: State variable ends behind valid data, label
+ */
+static const uint32_t stridesize_legacy = 0x53;
+
 /**
  * state_storage_write - Writes the given data to the storage
  * @param storage Storage object
@@ -173,17 +207,50 @@ int state_storage_read(struct state_backend_storage *storage,
 			zerobuckets++;
 		else if (!ret && !bucket_used)
 			bucket_used = bucket;
-		else if (ret)
-			dev_info(storage->dev, "Ignoring broken bucket %d@0x%08llx...\n", bucket->num, (long long) bucket->offset);
+		else if (ret) {
+			dev_notice(storage->dev, "Adjusting broken bucket %d@0x%08llx stridesize to legacy value 0x%x\n", bucket->num, (long long) bucket->offset, stridesize_legacy);
+			bucket->adjust(bucket, storage->offset + bucket->num * stridesize_legacy, stridesize_legacy);
+
+			ret = bucket->read(bucket, &bucket->buf, &bucket->len);
+			if (ret == -EUCLEAN)
+				bucket->needs_refresh = 1;
+			else if (ret)
+				continue;
+
+			ret = format->verify(format, magic, bucket->buf, &bucket->len, flags);
+			if (!ret && !bucket_used)
+				bucket_used = bucket;
+			if (ret) {
+				dev_notice(storage->dev, "Adjusting broken bucket %d@0x%08llx stridesize back to new value 0x%x\n", bucket->num, (long long) bucket->offset, storage->stridesize);
+				bucket->adjust(bucket, storage->offset + bucket->num * storage->stridesize, storage->stridesize);
+
+				ret = bucket->read(bucket, &bucket->buf, &bucket->len);
+				if (ret == -EUCLEAN)
+					bucket->needs_refresh = 1;
+				else if (ret)
+					continue;
+
+				ret = format->verify(format, magic, bucket->buf, &bucket->len, flags);
+				if (!ret && !bucket_used)
+					bucket_used = bucket;
+				if (ret)
+					dev_info(storage->dev, "Ignoring broken bucket %d@0x%08llx...\n", bucket->num, (long long) bucket->offset);
+			} else {
+				if (storage->stridesize != stridesize_legacy) {
+					dev_notice(storage->dev, "Adjusting storage stridesize to legacy value 0x%x\n", stridesize_legacy);
+					storage->stridesize = stridesize_legacy;
+				}
+			}
+		}
 	}
 
 	dev_dbg(storage->dev, "Checking redundant buckets finished.\n");
 
 	if (!bucket_used)
 		return dev_err_state_init(storage->dev, zerobuckets == totalbuckets ? -ENOMEDIUM : -ENOENT,
-					  "no valid state copy in any bucket\n");
+					  "no valid state copy in any bucket, stridesize value 0x%x\n", storage->stridesize);
 
-	dev_info(storage->dev, "Using bucket %d@0x%08llx\n", bucket_used->num, (long long) bucket_used->offset);
+	dev_info(storage->dev, "Using bucket %d@0x%08llxc, stridesize value 0x%x\n", bucket_used->num, (long long) bucket_used->offset, storage->stridesize);
 
 	/*
 	 * Restore/refresh all buckets except the one we currently use (in case
diff --git a/common/state/state.c b/common/state/state.c
index 8c34ae83e5..78faefd94a 100644
--- a/common/state/state.c
+++ b/common/state/state.c
@@ -355,7 +355,16 @@ struct device_node *state_to_node(struct state *state,
 		goto out;
 
 	for_each_child_of_node(state_root, child) {
-		ret = state_convert_node_variable(state, child, root, "", conv);
+		if ((state->storage.stridesize == 0x53) &&
+			(conv == STATE_CONVERT_FIXUP) &&
+			((!strcmp("update" , child->name)) ||
+			 (!strcmp("barebox", child->name)) ||
+			 (!strcmp("uart8"  , child->name)) ||
+			 (!strcmp("label"  , child->name)))) {
+			dev_notice(&state->dev, "Skipping %s for converting to fixup\n", child->name);
+		} else {
+			ret = state_convert_node_variable(state, child, root, "", conv);
+		}
 		if (ret)
 			goto out;
 	}
diff --git a/common/state/state.h b/common/state/state.h
index 0545cf6ac1..329403beda 100644
--- a/common/state/state.h
+++ b/common/state/state.h
@@ -39,6 +39,8 @@ struct state_backend_storage_bucket {
 		      const void * buf, ssize_t len);
 	int (*read) (struct state_backend_storage_bucket * bucket,
 		     void ** buf, ssize_t * len_hint);
+	void (*adjust) (struct state_backend_storage_bucket * bucket,
+		     off_t offset, ssize_t max_size);
 	void (*free) (struct state_backend_storage_bucket * bucket);
 
 	int num;
