diff --git a/arch/arm/dts/imx6ul-phytec-phycore-som-nand.dts b/arch/arm/dts/imx6ul-phytec-phycore-som-nand.dts
index 20c03184f..63bade301 100644
--- a/arch/arm/dts/imx6ul-phytec-phycore-som-nand.dts
+++ b/arch/arm/dts/imx6ul-phytec-phycore-som-nand.dts
@@ -52,3 +52,19 @@
 &usbotg2 {
 	status = "okay";
 };
+
+&pwm5 {
+	status = "okay";
+};
+
+&background {
+	status = "okay";
+};
+
+&i2c2 {
+	status = "okay";
+};
+
+&i2c3 {
+	status = "okay";
+};
diff --git a/arch/arm/dts/imx6ul-phytec-phycore-som.dtsi b/arch/arm/dts/imx6ul-phytec-phycore-som.dtsi
index 50b9469f2..42c20615c 100644
--- a/arch/arm/dts/imx6ul-phytec-phycore-som.dtsi
+++ b/arch/arm/dts/imx6ul-phytec-phycore-som.dtsi
@@ -25,13 +25,51 @@
 			device-path = &usdhc1, "partname:barebox-environment";
 			status = "disabled";
 		};
+	};
 
-		environment-sd2 {
-			compatible = "barebox,environment";
-			device-path = &usdhc2, "partname:barebox-environment";
-			status = "disabled";
+	regulators {
+		compatible = "simple-bus";
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		reg_pwm_3v3: regulator@1 {
+			compatible = "regulator-fixed";
+			reg = <2>;
+			regulator-name = "pwm-3v3";
+			regulator-min-microvolt = <3300000>;
+			regulator-max-microvolt = <3300000>;
+			status = "okay";
 		};
 	};
+
+	background: background {
+		compatible = "pwm-backlight";
+		/* PWM period is vriten in nanoseconds
+		   10KHz is 100.000 nanoseconds */
+		pwms = <&pwm5 0 100000>;
+		/* Brightnes is equal to PWM duty cycle */
+		brightness-levels = < 0  1  2  3  4  5  6  7  8  9
+							 10 11 12 13 14 15 16 17 18 19
+							 20 21 22 23 24 25 26 27 28 29
+							 30 31 32 33 34 35 36 37 38 39
+							 40 41 42 43 44 45 46 47 48 49
+							 50 51 52 53 54 55 56 57 58 59
+							 60 61 62 63 64 65 66 67 68 69
+							 70 71 72 73 74 75 76 77 78 79
+							 80 81 82 83 84 85 86 87 88 89
+							 90 91 92 93 94 95 96 97 98 99
+							 100>;
+		default-brightness-level = <50>;
+		power-supply = <&reg_pwm_3v3>;
+		status = "disabled";
+	};
+};
+
+
+&pwm5 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_pwm5>;
+	status = "disabled";
 };
 
 &fec1 {
@@ -82,29 +120,44 @@
 	clock-frequency = <100000>;
 	status = "disabled";
 
-	eeprom: eeprom@52 {
+	eeprom: eeprom@52 { /* EEPROM identification page (32 byte)*/
 		compatible = "cat,24c32";
 		reg = <0x52>;
 	};
 };
 
-&uart1 {
+&i2c2 {
 	pinctrl-names = "default";
-	pinctrl-0 = <&pinctrl_uart1>;
+	pinctrl-0 = <&pinctrl_i2c2>;
+	clock-frequency = <100000>;
 	status = "disabled";
 };
 
-&usbotg1 {
+&i2c3 {
 	pinctrl-names = "default";
-	pinctrl-0 = <&pinctrl_usb_otg1>;
-	dr_mode = "otg";
-	disable-over-current;
+	pinctrl-0 = <&pinctrl_i2c3>;
+	clock-frequency = <100000>;
 	status = "disabled";
+
+	gpio@20 {
+		compatible = "ti,tca6416";
+		gpio-controller;
+		#gpio-cells = <2>;
+		reg = <0x20>;
+	};
 };
 
-&usbotg2 {
-	dr_mode = "host";
-	disable-over-current;
+&ecspi1  {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_ecspi1>;
+	num-cs = <1>;
+	cs-gpios = <&gpio3 26 GPIO_ACTIVE_LOW>;
+	status = "okay";
+};
+
+&uart1 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_uart1>;
 	status = "disabled";
 };
 
@@ -128,100 +181,122 @@
 	};
 };
 
-&usdhc2 {
+&usbotg1 {
 	pinctrl-names = "default";
-	pinctrl-0 = <&pinctrl_usdhc2>;
-	bus-width = <8>;
-	non-removable;
-	status = "disabled";
+	pinctrl-0 = <&pinctrl_usb_otg1_id>;
+	dr_mode = "otg";
+	disable-over-current;
+	status = "okay";
+};
 
-	#address-cells = <1>;
-	#size-cells = <1>;
+&iomuxc {
+	pinctrl-name = "default";
 
-	partition@0 {
-		label = "barebox";
-		reg = <0x0 0xe0000>;
+	imx6ul-phytec-phycore-som {
+
+	pinctrl_enet1: enet1grp {
+		fsl,pins = <
+			MX6UL_PAD_GPIO1_IO07__ENET1_MDC			0x1b0b0
+			MX6UL_PAD_GPIO1_IO06__ENET1_MDIO		0x1b0b0
+			MX6UL_PAD_ENET1_RX_EN__ENET1_RX_EN		0x1b0b0
+			MX6UL_PAD_ENET1_RX_ER__ENET1_RX_ER		0x1b0b0
+			MX6UL_PAD_ENET1_RX_DATA0__ENET1_RDATA00	0x1b0b0
+			MX6UL_PAD_ENET1_RX_DATA1__ENET1_RDATA01	0x1b0b0
+			MX6UL_PAD_ENET1_TX_EN__ENET1_TX_EN		0x1b0b0
+			MX6UL_PAD_ENET1_TX_DATA0__ENET1_TDATA00	0x1b0b0
+			MX6UL_PAD_ENET1_TX_DATA1__ENET1_TDATA01	0x1b0b0
+			MX6UL_PAD_ENET1_TX_CLK__ENET1_REF_CLK1	0x4001b031
+		>;
 	};
 
-	partition@e0000 {
-		label = "barebox-environment";
-		reg = <0xe0000 0x20000>;
+	pinctrl_gpmi_nand: gpminandgrp {
+		fsl,pins = <
+			MX6UL_PAD_NAND_CLE__RAWNAND_CLE			0x0b0b1
+			MX6UL_PAD_NAND_ALE__RAWNAND_ALE			0x0b0b1
+			MX6UL_PAD_NAND_WP_B__RAWNAND_WP_B		0x0b0b1
+			MX6UL_PAD_NAND_READY_B__RAWNAND_READY_B	0x0b000
+			MX6UL_PAD_NAND_CE0_B__RAWNAND_CE0_B		0x0b0b1
+			MX6UL_PAD_NAND_RE_B__RAWNAND_RE_B		0x0b0b1
+			MX6UL_PAD_NAND_WE_B__RAWNAND_WE_B		0x0b0b1
+			MX6UL_PAD_NAND_DATA00__RAWNAND_DATA00	0x0b0b1
+			MX6UL_PAD_NAND_DATA01__RAWNAND_DATA01	0x0b0b1
+			MX6UL_PAD_NAND_DATA02__RAWNAND_DATA02	0x0b0b1
+			MX6UL_PAD_NAND_DATA03__RAWNAND_DATA03	0x0b0b1
+			MX6UL_PAD_NAND_DATA04__RAWNAND_DATA04	0x0b0b1
+			MX6UL_PAD_NAND_DATA05__RAWNAND_DATA05	0x0b0b1
+			MX6UL_PAD_NAND_DATA06__RAWNAND_DATA06	0x0b0b1
+			MX6UL_PAD_NAND_DATA07__RAWNAND_DATA07	0x0b0b1
+		>;
 	};
-};
 
-&iomuxc {
-        pinctrl-names = "default";
+	pinctrl_i2c1: i2c1grp {
+		fsl,pins = <
+			MX6UL_PAD_UART4_TX_DATA__I2C1_SCL	0x4001b8b0
+			MX6UL_PAD_UART4_RX_DATA__I2C1_SDA	0x4001b8b0
+		>;
+	};
 
-	imx6ul-phytec-phycore-som {
+	pinctrl_i2c2: i2c2grp {
+		fsl,pins = <
+			MX6UL_PAD_CSI_HSYNC__I2C2_SCL	0x4001b8b0
+			MX6UL_PAD_CSI_VSYNC__I2C2_SDA	0x4001b8b0
+		>;
+	};
 
-		pinctrl_enet1: enet1grp {
-			fsl,pins = <
-				MX6UL_PAD_GPIO1_IO07__ENET1_MDC		0x10010
-				MX6UL_PAD_GPIO1_IO06__ENET1_MDIO	0x10010
-				MX6UL_PAD_ENET1_RX_EN__ENET1_RX_EN	0x1b0b0
-				MX6UL_PAD_ENET1_RX_ER__ENET1_RX_ER	0x1b0b0
-				MX6UL_PAD_ENET1_RX_DATA0__ENET1_RDATA00	0x1b0b0
-				MX6UL_PAD_ENET1_RX_DATA1__ENET1_RDATA01	0x1b0b0
-				MX6UL_PAD_ENET1_TX_EN__ENET1_TX_EN	0x1b010
-				MX6UL_PAD_ENET1_TX_DATA0__ENET1_TDATA00	0x1b010
-				MX6UL_PAD_ENET1_TX_DATA1__ENET1_TDATA01	0x1b010
-				MX6UL_PAD_ENET1_TX_CLK__ENET1_REF_CLK1	0x4001b010
-			>;
-		};
+	pinctrl_i2c3: i2c3grp {
+		fsl,pins = <
+			MX6UL_PAD_ENET2_RX_DATA0__I2C3_SCL	0x4001b8b0
+			MX6UL_PAD_ENET2_RX_DATA1__I2C3_SDA	0x4001b8b0
+		>;
+	};
 
-		pinctrl_gpmi_nand: gpminandgrp {
-			fsl,pins = <
-				MX6UL_PAD_NAND_CLE__RAWNAND_CLE		0x0b0b1
-				MX6UL_PAD_NAND_ALE__RAWNAND_ALE		0x0b0b1
-				MX6UL_PAD_NAND_WP_B__RAWNAND_WP_B	0x0b0b1
-				MX6UL_PAD_NAND_READY_B__RAWNAND_READY_B	0x0b000
-				MX6UL_PAD_NAND_CE0_B__RAWNAND_CE0_B	0x0b0b1
-				MX6UL_PAD_NAND_RE_B__RAWNAND_RE_B	0x0b0b1
-				MX6UL_PAD_NAND_WE_B__RAWNAND_WE_B	0x0b0b1
-				MX6UL_PAD_NAND_DATA00__RAWNAND_DATA00	0x0b0b1
-				MX6UL_PAD_NAND_DATA01__RAWNAND_DATA01	0x0b0b1
-				MX6UL_PAD_NAND_DATA02__RAWNAND_DATA02	0x0b0b1
-				MX6UL_PAD_NAND_DATA03__RAWNAND_DATA03	0x0b0b1
-				MX6UL_PAD_NAND_DATA04__RAWNAND_DATA04	0x0b0b1
-				MX6UL_PAD_NAND_DATA05__RAWNAND_DATA05	0x0b0b1
-				MX6UL_PAD_NAND_DATA06__RAWNAND_DATA06	0x0b0b1
-				MX6UL_PAD_NAND_DATA07__RAWNAND_DATA07	0x0b0b1
-			>;
-		};
+	pinctrl_ecspi1: ecspi1grp {
+		fsl,pins = <
+			MX6UL_PAD_LCD_DATA20__ECSPI1_SCLK	0x1b0b1
+			MX6UL_PAD_LCD_DATA22__ECSPI1_MOSI	0x1b0b1
+			MX6UL_PAD_LCD_DATA23__ECSPI1_MISO	0x1b0b1
+			MX6UL_PAD_LCD_DATA21__ECSPI1_SS0	0x80000000
+		>;
+	};
 
-		pinctrl_i2c1: i2cgrp {
-			fsl,pins = <
-				MX6UL_PAD_UART4_TX_DATA__I2C1_SCL	0x4001b8b0
-				MX6UL_PAD_UART4_RX_DATA__I2C1_SDA	0x4001b8b0
-			>;
-		};
+	pinctrl_uart1: uart1grp {
+		fsl,pins = <
+			MX6UL_PAD_UART1_TX_DATA__UART1_DCE_TX	0x1b0b1
+			MX6UL_PAD_UART1_RX_DATA__UART1_DCE_RX	0x1b0b1
+		>;
+	};
 
-		pinctrl_uart1: uart1grp {
-			fsl,pins = <
-				MX6UL_PAD_UART1_TX_DATA__UART1_DCE_TX	0x1b0b1
-				MX6UL_PAD_UART1_RX_DATA__UART1_DCE_RX	0x1b0b1
-			>;
-		};
+	pinctrl_usdhc1: usdhc1grp {
+		fsl,pins = <
+			MX6UL_PAD_SD1_CMD__USDHC1_CMD		0x17059
+			MX6UL_PAD_SD1_CLK__USDHC1_CLK		0x10059
+			MX6UL_PAD_SD1_DATA0__USDHC1_DATA0	0x17059
+			MX6UL_PAD_SD1_DATA1__USDHC1_DATA1	0x17059
+			MX6UL_PAD_SD1_DATA2__USDHC1_DATA2	0x17059
+			MX6UL_PAD_SD1_DATA3__USDHC1_DATA3	0x17059
+			MX6UL_PAD_UART1_RTS_B__GPIO1_IO19	0x17059 /* SD1 CD */
+		>;
+	};
 
-		pinctrl_usb_otg1: usbotg1grp {
-			fsl,pins = <
-				MX6UL_PAD_GPIO1_IO00__ANATOP_OTG1_ID	0x17059
-			>;
-		};
+	pinctrl_usb_otg1_id: usbotg1idgrp {
+		fsl,pins = <
+			MX6UL_PAD_GPIO1_IO00__ANATOP_OTG1_ID	0x17059
+		>;
+	};
 
-		pinctrl_usdhc1: usdhc1grp {
-			fsl,pins = <
-				MX6UL_PAD_SD1_CMD__USDHC1_CMD     	0x17059
-				MX6UL_PAD_SD1_CLK__USDHC1_CLK     	0x10059
-				MX6UL_PAD_SD1_DATA0__USDHC1_DATA0 	0x17059
-				MX6UL_PAD_SD1_DATA1__USDHC1_DATA1 	0x17059
-				MX6UL_PAD_SD1_DATA2__USDHC1_DATA2 	0x17059
-				MX6UL_PAD_SD1_DATA3__USDHC1_DATA3 	0x17059
-				MX6UL_PAD_UART1_RTS_B__GPIO1_IO19       0x17059 /* SD1 CD */
-			>;
-		};
+	pinctrl_io: iogrp {
+		fsl,pins = <
+				 MX6UL_PAD_UART2_RTS_B__GPIO1_IO23 0x79 /* Force OTG1 ID */
+		>;
+	};
 
-		pinctrl_usdhc2: usdhc2grp {
+	pinctrl_pwm5: pwm5grp {
+		fsl,pins = <
+			MX6UL_PAD_LCD_DATA18__PWM5_OUT	0x1b0b0
+		>;
+	};
+
+	pinctrl_usdhc2: usdhc2grp {
 			fsl,pins = <
 				MX6UL_PAD_NAND_WE_B__USDHC2_CMD     	0x17059
 				MX6UL_PAD_NAND_RE_B__USDHC2_CLK     	0x10059
@@ -234,7 +309,7 @@
 				MX6UL_PAD_NAND_DATA06__USDHC2_DATA6 	0x17059
 				MX6UL_PAD_NAND_DATA07__USDHC2_DATA7 	0x17059
 			>;
-		};
+	};
 	};
 };
 
diff --git a/arch/arm/dts/imx6ul-phytec-state.dtsi b/arch/arm/dts/imx6ul-phytec-state.dtsi
index a21cb49f0..f141c933d 100644
--- a/arch/arm/dts/imx6ul-phytec-state.dtsi
+++ b/arch/arm/dts/imx6ul-phytec-state.dtsi
@@ -9,73 +9,102 @@
 		state = &state;
 	};
 
-	state: imx6ul_phytec_boot_state {
-		magic = <0x883b86a6>;
+	state: boot_states {
+		magic = <0x27031977>;
 		compatible = "barebox,state";
 		backend-type = "raw";
-		backend = <&backend_update_eeprom>;
-		backend-stridesize = <54>;
+		/* backend = <&backend_state_eeprom>; */
+		backend = <&eeprom>;
+		/* backend = &eeprom; */
+		backend-storage-type = "direct";
+		backend-stridesize = <0x70>;
 		status = "disabled";
 
-		#address-cells = <1>;
-		#size-cells = <1>;
-		bootstate {
-			#address-cells = <1>;
-			#size-cells = <1>;
-			last_chosen {
-				reg = <0x0 0x4>;
-				type = "uint32";
-			};
-			system0 {
-				#address-cells = <1>;
-				#size-cells = <1>;
-				remaining_attempts {
-					reg = <0x4 0x4>;
-					type = "uint32";
-					default = <3>;
-				};
-				priority {
-					reg = <0x8 0x4>;
-					type = "uint32";
-					default = <21>;
-				};
-				ok {
-					reg = <0xc 0x4>;
-					type = "uint32";
-					default = <0>;
-				};
-			};
-			system1 {
-				#address-cells = <1>;
-				#size-cells = <1>;
-				remaining_attempts {
-					reg = <0x10 0x4>;
-					type = "uint32";
-					default = <3>;
-				};
-				priority {
-					reg = <0x14 0x4>;
-					type = "uint32";
-					default = <20>;
-				};
-				ok {
-					reg = <0x18 0x4>;
-					type = "uint32";
-					default = <0>;
-				};
-			};
+		/* #address-cells = <1>; */
+		/* #size-cells = <1>; */
+		locktheft {
+			reg = <0x00 0x1>;
+			type = "uint8";
+			default = <0x0>;
+		};
+
+		partition {
+			reg = <0x01 0x4>;
+			type = "enum32";
+			names = "0", "1";
+			default = <0x0>;
+		};
+
+		order {
+			reg = <0x05 0x9>;
+			type = "string";
+			default = "000000000";
+		};
+
+		serial {
+			reg = <0x0e 0x6>;
+			type = "string";
+			default = "000000";
+		};
+
+		item {
+			reg = <0x14 0x7>;
+			type = "string";
+			default = "0000000";
+		};
+
+		revision {
+			reg = <0x1b 0x3>;
+			type = "string";
+			default = "000";
+		};
+
+		date {
+			reg = <0x1e 0x6>;
+			type = "string";
+			default = "000000";
+		};
+
+		update {
+			reg = <0x24 0x4>;
+			type = "enum32";
+			names = "0", "1";
+			default = <0x0>;
+		};
+
+		barebox {
+			reg = <0x28 0x4>;
+			type = "enum32";
+			names = "0", "1";
+			default = <0x0>;
+		};
+
+		uart8 {
+			reg = <0x2c 0x4>;
+			type = "enum32";
+			names = "0", "1";
+			default = <0x1>;
+		};
+
+		label {
+			reg = <0x30 0x14>;
+			type = "string";
+			default = "000000-0000000-00000";
 		};
 	};
 };
 
+#if 0
 &eeprom {
 	partitions {
 		compatible = "fixed-partitions";
 		#size-cells = <1>;
 		#address-cells = <1>;
-		backend_update_eeprom: state@0 {
+		backend_state_eeprom: state@0 {
+		/* backend_state_eeprom { */
 			reg = <0x0 0x100>;
-			label = "update-eeprom";
+			label = "boot-eeprom";
 		};
 	};
 };
+#endif
diff --git a/defaultenv/defaultenv-2-base/bin/init b/defaultenv/defaultenv-2-base/bin/init
index 8d02e3d3a..2a71bc90c 100644
--- a/defaultenv/defaultenv-2-base/bin/init
+++ b/defaultenv/defaultenv-2-base/bin/init
@@ -1,5 +1,9 @@
 #!/bin/sh
 
+
+# set backlight to 50%
+backlight0.brightness=50
+
 export PATH=/env/bin
 
 global hostname
