diff --git a/defaultenv/defaultenv-2-base/bin/init b/defaultenv/defaultenv-2-base/bin/init
index 8d02e3d3a..d7940d0c3 100644
--- a/defaultenv/defaultenv-2-base/bin/init
+++ b/defaultenv/defaultenv-2-base/bin/init
@@ -21,7 +21,7 @@ magicvar -a global.user "username (used in network filenames)"
 magicvar -a global.autoboot_timeout "timeout in seconds before automatic booting"
 [ -z "${global.autoboot_abort_key}" ] && global.autoboot_abort_key=any
 magicvar -a global.autoboot_abort_key "key to abort automatic booting (valid options: any, ctrl-c)"
-[ -z "${global.boot.default}" ] && global.boot.default=net
+[ -z "${global.boot.default}" ] && global.boot.default=update
 [ -z "${global.editcmd}" ] && global.editcmd=sedit
 
 [ -e /env/config-board ] && /env/config-board
