diff --git a/common/boot.c b/common/boot.c
index f546fce62c..7a96030c0a 100644
--- a/common/boot.c
+++ b/common/boot.c
@@ -124,7 +124,7 @@ static char *global_user;
 
 static int init_boot(void)
 {
-	global_boot_default = xstrdup("net");
+	global_boot_default = xstrdup("update");
 	globalvar_add_simple_string("boot.default", &global_boot_default);
 	globalvar_add_simple_int("boot.watchdog_timeout",
 				 &boot_watchdog_timeout, "%u");
