diff --git a/arch/arm/boards/phytec-som-imx6/board.c b/arch/arm/boards/phytec-som-imx6/board.c
index 1e515a093a..7985623620 100644
--- a/arch/arm/boards/phytec-som-imx6/board.c
+++ b/arch/arm/boards/phytec-som-imx6/board.c
@@ -202,7 +202,22 @@ static int physom_imx6_probe(struct device_d *dev)
 						"/dev/mmc0", 0);
 
 	} else if (flags & IS_PHYCORE_UL) {
-		barebox_set_hostname("phyCORE-i.MX6UL");
+		ret = of_devices_ensure_probed_by_property("gpio-controller");
+		if (ret)
+			return ret;
+
+		gpio_direction_output(IMX_GPIO_NR(5, 1), 1);
+		pr_notice("powerlock: enabled\n"); // P_L_SIGNAL on gpio129
+
+		int cpu_type = __imx6_cpu_type();
+		if (cpu_type == IMX6_CPUTYPE_IMX6UL) {
+			barebox_set_hostname("imx6ul-medusa");
+		} else if (cpu_type == IMX6_CPUTYPE_IMX6ULL) {
+			barebox_set_hostname("imx6ull-medusa");
+		} else {
+			pr_err("Determining CPU type failed\n");
+		}
+		pr_notice("Set hostname to %s \n", barebox_get_hostname());
 		default_environment_path = "/chosen/environment-nand";
 		default_envdev = "NAND flash";
 
